// ng2-google-charts/esm2015/lib/google-charts-control/google-charts-control.component.js
import { __awaiter, __decorate } from "tslib";
import { Component, OnInit, Input, ElementRef } from '@angular/core';
import { GoogleChartsLoaderService } from '../google-charts-loader.service';
let GoogleChartsControlComponent = class GoogleChartsControlComponent {
    constructor(el, loaderService) {
        this.el = el;
        this.loaderService = loaderService;
    }
    ngOnInit() {
        this.data.component = this;
    }
    ensureInit() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.wrapper) {
                return;
            }
            yield this.loaderService.load({ packages: ['controls'] });
            let opt;
            opt = Object.create(this.data);
            opt.containerId = this.el.nativeElement.querySelector('div');
            this.wrapper = new google.visualization.ControlWrapper(opt);
        });
    }
};
GoogleChartsControlComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: GoogleChartsLoaderService }
];
__decorate([
    Input()
], GoogleChartsControlComponent.prototype, "data", void 0);
GoogleChartsControlComponent = __decorate([
    Component({
        selector: 'google-charts-control',
        template: '<div></div>'
    })
], GoogleChartsControlComponent);
export { GoogleChartsControlComponent };
